prov/util/src/util_cq.c

```diff
@@ -258,6 +258,10 @@ ssize_t ofi_cq_readfrom(struct fid_cq *cq_fid, void *buf, size_t count,
 	cq = container_of(cq_fid, struct util_cq, cq_fid);
 
 	cq->progress(cq);
+
+	if (cq->flags & FI_PEER)
+		return 0;
```

Comment:
Is there not an ofi_cq_peer_read() call that simply calls progress that can be used?  If not, having one would be preferred to inserting this check in the fast path of all cq read calls.

Reply:
Agreed with Sean. I would prefer not to have this added check in the regular path. The definition of the peer implementation is that you can only drive progress on it by calling fi_cq_read with NULL and 0 to it never tries to read the CQ. We should set the FI_PEER cq_read function pointer so it only drives progress.

Reply:
I see, that makes sense to me, currently the util_peer_cq_ops has cq_read as util_cq_read, which is ofi_cq_readfrom. I think I just need to introduce a ofi_cq_peer_read() that only progress the cq. 

```
static ssize_t util_cq_read(struct fid_cq *cq_fid, void *buf, size_t count)
{
       return ofi_cq_readfrom(cq_fid, buf, count, NULL);
}

static struct fi_ops_cq util_peer_cq_ops = {
       .size = sizeof(struct fi_ops_cq),
       .read = util_cq_read,
       .readfrom = fi_no_cq_readfrom,
       .readerr = fi_no_cq_readerr,
       .sread = fi_no_cq_sread,
       .sreadfrom = fi_no_cq_sreadfrom,
       .signal = fi_no_cq_signal,
       .strerror = fi_no_cq_strerror,
};
```

Is that right?

Reply:
@shijin-aws Yes, I think that's better. Thanks for catching!

Reply:
Thanks. `util_cq_read` becomes unused after I introduce `ofi_peer_cq_read` so I removed it. Question: Is `ofi_*` function aimed to be non-static?

Reply:
I think this is good, and I wouldn't worry about ofi vs util as the prefix.  I'd rather see ofi myself.

Reply:
I believe ofi_* functions are meant to be called from providers and util_* functions called internally within the util code. That's at least what I've been doing by @shefty can confirm or contradict

Reply:
Thank you both. Rephrased my commit message to reflect the updated change.

Reply:
I believe that's the case in general.  I.e. util_ are static.  But I don't think that distinction is particularly useful, so I wouldn't worry about it.